subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -266,7 +268,60 @@ static int hfp_ag_next_step(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *us
 
 static int hfp_ag_send(struct bt_hfp_ag *ag, struct bt_ag_tx *tx)
 {
-	return bt_rfcomm_dlc_send(&ag->rfcomm_dlc, tx->buf);
+	int err = bt_rfcomm_dlc_send(&ag->rfcomm_dlc, tx->buf);
+
+	if (err < 0) {
+		net_buf_unref(tx->buf);
+	}
+
+	return err;
+}
+
+static void bt_ag_tx_work(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct bt_hfp_ag *ag = CONTAINER_OF(dwork, struct bt_hfp_ag, tx_work);
+	sys_snode_t *node;
+	struct bt_ag_tx *tx;
+	bt_hfp_state_t state;
+
+	hfp_ag_lock(ag);
+	state = ag->state;
+	hfp_ag_unlock(ag);
+	if ((state == BT_HFP_DISCONNECTED) || (state == BT_HFP_DISCONNECTING)) {
+		LOG_ERR("AG %p is not connected", ag);
+		return;
+	}
+
+	hfp_ag_lock(ag);
+	node = sys_slist_peek_head(&ag->tx_pending);
+	hfp_ag_unlock(ag);
+	if (!node) {
+		LOG_DBG("No pending tx");
+		return;
+	}
+
+	tx = CONTAINER_OF(node, struct bt_ag_tx, node);
```

Comment:
Since you're not holding the lock anymore at this point, doesn't it mean that the list could have changed and the data invalidated. It seems to me like you should be holding the lock until you're completely done with accessing the `ag` pointer in this function.

Reply:
Updated.